File: src/leetcode/P038_CountAndSay.java
package leetcode;

public class P038_CountAndSay {
	/*
	 * 	2 ms
	 * 	85.40%
	 */
	static class Solution {
		public String countAndSay(int n) {
			if (n < 1) {
				return null;
			}
			String result = "1";
			for (int i = 2; i <= n; i++) {
				result = countAndSay(result);
			}
			return result;
		}
		public String countAndSay(String str) {
			StringBuilder builder = new StringBuilder(128);
			int count = 1;
			for (int i = 1; i < str.length(); i++) {
				if (str.charAt(i) == str.charAt(i - 1)) {
					count++;
				} else {
					builder.append(count);
					builder.append(str.charAt(i - 1));
					count = 1;
				}
			}
			builder.append(count);
			builder.append(str.charAt(str.length() - 1));
			return builder.toString();
		}
	}
}

File: python/P038_CountAndSay.py
#coding=utf-8

'''
    url: leetcode.com/problems/count-and-say/
    @author:     zxwtry
    @email:      zxwtry@qq.com
    @date:       2017年4月5日
    @details:    Solution: 72ms 13.30%
'''

class Solution(object):
    def countAndSay(self, n):
        """
        :type n: int
        :rtype: str
        """
        if n < 1: return ""
        if n == 1: return "1"
        answer="1"
        for i in range(n-1):
            record, pre_char, pre_count=[], "ddd", 0
            for j in range(len(answer)):
                if answer[j] == pre_char:
                    pre_count += 1
                else:
                    if pre_count != 0:
                        record.append(str(pre_count))
                        record.append(pre_char)
                    pre_char = answer[j]
                    pre_count = 1
            if pre_count != 0:
                record.append(str(pre_count))
                record.append(pre_char)
            answer = "".join(record)
        return answer

if __name__ == "__main__":
    print(Solution().countAndSay(4))
                    
                    
                